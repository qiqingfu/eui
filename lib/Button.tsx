import React from 'react';

const Button = (): React.ReactElement => {
  return (
    <>
      <button>click</button>
    </>
  );
};

export default Button;
