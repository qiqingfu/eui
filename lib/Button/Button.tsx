import React, { useMemo } from 'react';
import classNames from 'classnames';
import { Icon } from '../index';

export type ButtonSize = 'small' | 'middle' | 'large';
interface ButtonProps extends React.DOMAttributes<Element> {
  size?: ButtonSize;
  disabled?: boolean;
  loading?: boolean;
  // https://github.com/yannickcr/eslint-plugin-react/issues/7#issuecomment-90294004
  children: React.ReactNode;
}

const Button: React.FC<ButtonProps> = (props: ButtonProps): React.ReactElement => {
  const { size = 'middle', loading = false, children, ...restProps } = props;
  const { disabled } = restProps;

  const isLoading = useMemo(() => loading && !disabled, [loading, disabled]);

  const cls = classNames('e-btn', size, { 'e-btn_loading': isLoading });

  return (
    <button {...restProps} className={cls}>
      {isLoading && <Icon name="loading" className="e-btn__icon" />}
      <span>{children}</span>
    </button>
  );
};

export default Button;
