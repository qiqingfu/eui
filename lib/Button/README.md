## 需求分析

### size

- 支持三种程度的尺寸
- 默认 middle
- 大尺寸 large
- 中等尺寸 middle
- 小尺寸 small
