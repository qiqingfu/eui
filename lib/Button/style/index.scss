@import "../../style/var";

.#{$e-prefix}-btn {
  display: inline-block;
  background-color: $primay;
  border: none;
  outline: none;
  color: #fff;
  font-size: 14px;
  cursor: pointer;
  border-radius: 4px;
  transition: all .3s cubic-bezier(.645,.045,.355,1);

  &.small {
    padding: 9px 15px;
    border-radius: 3px;
  }

  &.middle {
    padding: 10px 20px;
  }

  &.large {
    padding: 12px 20px;
  }

  &[disabled] {
    color: #00000040;
    border: solid 1px #d9d9d9;
    background-color: #f5f5f5;
    text-shadow: none;
    box-shadow: none;
    cursor: not-allowed;
  }

  &:not([disabled]):hover {
    background-color: #40a9ff;
    border-color: #40a9ff;
  }
}