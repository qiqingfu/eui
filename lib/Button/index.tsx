import Button from './Button';
export { ButtonSize } from './Button';

export default Button;
