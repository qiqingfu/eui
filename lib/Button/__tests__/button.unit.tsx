import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Button from '../index';

describe('Button Component', () => {
  it('Button renders correctly', () => {
    const tree = renderer.create(<Button>Button</Button>);
    expect(tree).toMatchSnapshot();
  });

  it('Button default children', () => {
    const value = 'Default Value';
    const wrapper = shallow(<Button>{value}</Button>);
    const result = wrapper.find('span').last();

    expect(result.text()).toBe(value);
  });

  describe('Button props.size', () => {
    it('Default size middle', () => {
      const wrapper = shallow(<Button>Button</Button>);

      expect(wrapper.hasClass('middle')).toBeTruthy();
    });

    it('Set the value of props.size to small ', () => {
      const size = 'small';
      const wrapper = shallow(<Button>Button</Button>);

      wrapper.setProps({ size });
      expect(wrapper.hasClass(size)).toBeTruthy();
    });
  });

  it('Button props.disabled', () => {
    // CSS Selector https://developer.mozilla.org/zh-CN/docs/Web/CSS/Attribute_selectors
    const wrapper = shallow(<Button>Disabled Button</Button>);
    wrapper.setProps({ disabled: false });
    expect(wrapper.is('[disabled=true]')).toBeFalsy();
    wrapper.setProps({ disabled: true });
    expect(wrapper.is('[disabled=true]')).toBeTruthy();
  });
});
