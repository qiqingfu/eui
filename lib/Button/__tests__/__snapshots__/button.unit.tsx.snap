// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Button renders correctly 1`] = `
<button
  className="__e-ui-btn middle"
>
  <span>
    Button
  </span>
</button>
`;
